File: backend/app/services/vapi_voice_service.py
"""
Production Vapi Voice Service
Secure integration with proper key management and enhanced TTS/STT
"""

import asyncio
import json
import logging
import base64
import tempfile
import os
from typing import Dict, Any, Optional, List
from datetime import datetime
import httpx
from fastapi.responses import StreamingResponse
from io import BytesIO

from app.core.config import settings

logger = logging.getLogger(__name__)

class VapiVoiceService:
    """Production-ready Vapi AI Voice Service with secure key management"""
    
    def __init__(self):
        # Secure key management - private key only on backend
        self.private_key = settings.VAPI_PRIVATE_KEY
        self.public_key = settings.VAPI_PUBLIC_KEY  # Safe for frontend
        self.assistant_id = settings.VAPI_ASSISTANT_ID
        self.phone_number = settings.VAPI_PHONE_NUMBER
        self.webhook_secret = settings.VAPI_WEBHOOK_SECRET
        self.base_url = settings.VAPI_BASE_URL
        
        # Enhanced voice configuration
        self.voice_config = {
            "provider": "openai",  # Using OpenAI for reliability
            "model": "tts-1",
            "voice": "alloy",  # Professional female voice
            "speed": 1.0,
            "response_format": "mp3"
        }
        
        # Financial assistant system message
        self.system_message = self._get_financial_system_message()
        
        # Multi-language support
        self.supported_languages = {
            "english": {"code": "en-IN", "voice": "alloy"},
            "hindi": {"code": "hi-IN", "voice": "nova"},
            "tamil": {"code": "ta-IN", "voice": "shimmer"},
            "telugu": {"code": "te-IN", "voice": "echo"},
            "bengali": {"code": "bn-IN", "voice": "fable"},
            "marathi": {"code": "mr-IN", "voice": "onyx"}
        }
        
        logger.info("Vapi Voice Service initialized with secure configuration")
    
    def get_public_config(self) -> Dict[str, Any]:
        """Get safe public configuration for frontend"""
        return {
            "public_key": self.public_key,  # Safe to expose
            "assistant_id": self.assistant_id,
            "supported_languages": list(self.supported_languages.keys()),
            "base_url": self.base_url,
            "features": {
                "voice_calls": True,
                "text_to_speech": True,
                "speech_to_text": True,
                "real_time_conversation": True,
                "financial_context": True,
                "multi_language": True
            },
            "voice_options": [voice["voice"] for voice in self.supported_languages.values()],
            "status": "active" if self.private_key else "inactive"
        }
    
    async def create_financial_assistant(
        self, 
        language: str = "english",
        user_context: Optional[Dict[str, Any]] = None
    ) -> Optional[str]:
        """Create specialized financial assistant"""
        
        if not self.private_key:
            logger.warning("Vapi private key not configured")
            return None
        
        try:
            # Enhanced system message with user context
            system_message = self.system_message
            if user_context:
                context_msg = self._build_context_message(user_context)
                system_message += f"\n\nCURRENT USER CONTEXT:\n{context_msg}"
            
            # Language-specific voice configuration
            lang_config = self.supported_languages.get(language, self.supported_languages["english"])
            
            assistant_config = {
                "name": f"FinVoice Assistant ({language.title()})",
                "firstMessage": self._get_greeting_message(language),
                "systemMessage": system_message,
                "model": {
                    "provider": "openai",
                    "model": "gpt-4",
                    "temperature": 0.7,
                    "maxTokens": 300,
                    "functions": self._get_financial_functions()
                },
                "voice": {
                    "provider": "openai",
                    "model": self.voice_config["model"],
                    "voice": lang_config["voice"],
                    "speed": self.voice_config["speed"]
                },
                "silenceTimeoutSeconds": 30,
                "maxDurationSeconds": 600,  # 10 minutes max
                "recordingEnabled": True,
                "endCallMessage": self._get_goodbye_message(language),
                "serverUrl": f"{settings.ALLOWED_ORIGINS[0]}/api/v1/vapi/webhook" if settings.ALLOWED_ORIGINS else None,
                "serverUrlSecret": self.webhook_secret
            }
            
            async with httpx.AsyncClient(timeout=30.0) as client:
                response = await client.post(
                    f"{self.base_url}/assistant",
                    headers={
                        "Authorization": f"Bearer {self.private_key}",
                        "Content-Type": "application/json"
                    },
                    json=assistant_config
                )
                
                if response.status_code in [200, 201]:
                    assistant_data = response.json()
                    assistant_id = assistant_data.get("id")
                    logger.info(f"Created Vapi assistant: {assistant_id}")
                    return assistant_id
                else:
                    logger.error(f"Assistant creation failed: {response.status_code} - {response.text}")
                    return None
                    
        except Exception as e:
            logger.error(f"Failed to create assistant: {e}")
            return None
    
    async def start_voice_call(
        self, 
        phone_number: str, 
        language: str = "english",
        user_context: Optional[Dict[str, Any]] = None
    ) -> Optional[str]:
        """Start voice call with financial context"""
        
        if not self.private_key or not self.phone_number:
            logger.error("Vapi configuration incomplete for voice calls")
            return None
        
        try:
            # Create or get assistant
            assistant_id = await self.create_financial_assistant(language, user_context)
            if not assistant_id:
                return None
            
            call_config = {
                "assistant": {"assistantId": assistant_id},
                "phoneNumberId": self.phone_number,
                "customer": {"number": phone_number},
                "metadata": {
                    "language": language,
                    "user_context": "financial_assistance",
                    "timestamp": datetime.now().isoformat()
                }
            }
            
            async with httpx.AsyncClient(timeout=30.0) as client:
                response = await client.post(
                    f"{self.base_url}/call",
                    headers={
                        "Authorization": f"Bearer {self.private_key}",
                        "Content-Type": "application/json"
                    },
                    json=call_config
                )
                
                if response.status_code == 200:
                    call_data = response.json()
                    call_id = call_data.get("id")
                    logger.info(f"Started Vapi call: {call_id}")
                    return call_id
                else:
                    logger.error(f"Call start failed: {response.status_code} - {response.text}")
                    return None
                    
        except Exception as e:
            logger.error(f"Error starting voice call: {e}")
            return None
    
    async def generate_speech_audio(
        self, 
        text: str, 
        language: str = "english",
        voice_type: str = "alloy"
    ) -> Optional[bytes]:
        """Generate high-quality speech audio using OpenAI TTS"""
        
        try:
            # Use OpenAI TTS for reliability
            if settings.OPENAI_API_KEY:
                return await self._generate_openai_tts(text, voice_type)
            else:
                # Fallback to mock audio for demo
                return await self._generate_mock_audio(text)
                
        except Exception as e:
            logger.error(f"Speech generation failed: {e}")
            return None
    
    async def _generate_openai_tts(self, text: str, voice: str = "alloy") -> Optional[bytes]:
        """Generate speech using OpenAI TTS API"""
        
        try:
            async with httpx.AsyncClient(timeout=30.0) as client:
                response = await client.post(
                    "https://api.openai.com/v1/audio/speech",
                    headers={
                        "Authorization": f"Bearer {settings.OPENAI_API_KEY}",
                        "Content-Type": "application/json"
                    },
                    json={
                        "model": "tts-1",
                        "input": text,
                        "voice": voice,
                        "response_format": "mp3",
                        "speed": 1.0
                    }
                )
                
                if response.status_code == 200:
                    return response.content
                else:
                    logger.error(f"OpenAI TTS failed: {response.status_code}")
                    return None
                    
        except Exception as e:
            logger.error(f"OpenAI TTS error: {e}")
            return None
    
    async def _generate_mock_audio(self, text: str) -> bytes:
        """Generate mock audio for demo purposes"""
        # Create a simple audio file placeholder
        # In production, you could use pyttsx3 or another local TTS
        mock_audio = b"RIFF" + b"\x00" * 44 + b"mock audio data for: " + text.encode()[:100]
        return mock_audio
    
    async def create_tts_stream(
        self, 
        text: str, 
        language: str = "english",
        voice_type: str = "alloy"
    ) -> StreamingResponse:
        """Create streaming TTS response for direct browser playback"""
        
        try:
            audio_data = await self.generate_speech_audio(text, language, voice_type)
            
            if audio_data:
                audio_stream = BytesIO(audio_data)
                
                return StreamingResponse(
                    iter([audio_data]),
                    media_type="audio/mpeg",
                    headers={
                        "Content-Disposition": "inline; filename=speech.mp3",
                        "Cache-Control": "no-cache",
                        "Accept-Ranges": "bytes"
                    }
                )
            else:
                # Return empty audio response
                return StreamingResponse(
                    iter([b""]),
                    media_type="audio/mpeg",
                    status_code=204
                )
                
        except Exception as e:
            logger.error(f"TTS streaming failed: {e}")
            return StreamingResponse(
                iter([b""]),
                media_type="audio/mpeg",
                status_code=500
            )
    
    async def process_voice_message(
        self, 
        message: str, 
        user_context: Dict[str, Any],
        language: str = "english"
    ) -> Dict[str, Any]:
        """Process voice message with financial context injection"""
        
        try:
            # Enhanced prompt with user context
            enhanced_prompt = self._build_contextual_prompt(message, user_context, language)
            
            # Generate AI response
            ai_response = await self._get_financial_ai_response(enhanced_prompt)
            
            # Generate speech audio
            audio_data = await self.generate_speech_audio(ai_response, language)
            
            return {
                "success": True,
                "text_response": ai_response,
                "audio_data": base64.b64encode(audio_data).decode() if audio_data else None,
                "language": language,
                "timestamp": datetime.now().isoformat(),
                "user_context_injected": True
            }
            
        except Exception as e:
            logger.error(f"Voice message processing failed: {e}")
            return {
                "success": False,
                "error": str(e),
                "text_response": "I'm sorry, I encountered an error processing your request.",
                "audio_data": None
            }
    
    def _build_contextual_prompt(
        self, 
        message: str, 
        user_context: Dict[str, Any], 
        language: str
    ) -> str:
        """Build enhanced prompt with user financial context"""
        
        context_summary = self._build_context_message(user_context)
        
        return f"""
        You are FinVoice AI, an expert financial advisor for Indian users.
        
        USER FINANCIAL CONTEXT:
        {context_summary}
        
        USER MESSAGE: {message}
        LANGUAGE: {language}
        
        Provide a helpful, specific financial response that:
        1. References their actual financial data when relevant
        2. Gives actionable advice
        3. Is conversational and empathetic
        4. Keeps response under 150 words
        5. Uses Indian Rupees (₹) for amounts
        
        Response:
        """
    
    def _build_context_message(self, user_context: Dict[str, Any]) -> str:
        """Build readable context message from user data"""
        
        context_parts = []
        
        if user_context.get("portfolio_value"):
            context_parts.append(f"Portfolio: ₹{user_context['portfolio_value']:,.0f}")
        
        if user_context.get("monthly_income"):
            context_parts.append(f"Monthly Income: ₹{user_context['monthly_income']:,.0f}")
            
        if user_context.get("monthly_expenses"):
            context_parts.append(f"Monthly Expenses: ₹{user_context['monthly_expenses']:,.0f}")
        
        if user_context.get("goals"):
            active_goals = len(user_context["goals"])
            context_parts.append(f"Active Goals: {active_goals}")
        
        if user_context.get("recent_transactions"):
            recent_count = len(user_context["recent_transactions"])
            context_parts.append(f"Recent Transactions: {recent_count}")
        
        return "; ".join(context_parts) if context_parts else "Basic financial profile"
    
    async def _get_financial_ai_response(self, prompt: str) -> str:
        """Get AI response for financial queries"""
        
        # Use OpenRouter or OpenAI for response generation
        try:
            if settings.OPENROUTER_API_KEY:
                return await self._get_openrouter_response(prompt)
            elif settings.OPENAI_API_KEY:
                return await self._get_openai_response(prompt)
            else:
                return self._get_fallback_response(prompt)
                
        except Exception as e:
            logger.error(f"AI response generation failed: {e}")
            return "I'm here to help with your financial questions. Could you please try asking again?"
    
    async def _get_openai_response(self, prompt: str) -> str:
        """Get response from OpenAI API"""
        
        try:
            async with httpx.AsyncClient(timeout=30.0) as client:
                response = await client.post(
                    "https://api.openai.com/v1/chat/completions",
                    headers={
                        "Authorization": f"Bearer {settings.OPENAI_API_KEY}",
                        "Content-Type": "application/json"
                    },
                    json={
                        "model": "gpt-3.5-turbo",
                        "messages": [
                            {"role": "system", "content": "You are FinVoice AI, a helpful financial advisor."},
                            {"role": "user", "content": prompt}
                        ],
                        "max_tokens": 200,
                        "temperature": 0.7
                    }
                )
                
                if response.status_code == 200:
                    data = response.json()
                    return data["choices"][0]["message"]["content"]
                else:
                    logger.error(f"OpenAI API failed: {response.status_code}")
                    return self._get_fallback_response(prompt)
                    
        except Exception as e:
            logger.error(f"OpenAI API error: {e}")
            return self._get_fallback_response(prompt)
    
    def _get_fallback_response(self, prompt: str) -> str:
        """Generate fallback response when AI services are unavailable"""
        
        prompt_lower = prompt.lower()
        
        if "portfolio" in prompt_lower:
            return "Your portfolio is performing well. Based on current market conditions, consider reviewing your asset allocation and rebalancing if needed. Would you like me to analyze specific holdings?"
        
        elif "expense" in prompt_lower or "spending" in prompt_lower:
            return "I can help you track and optimize your expenses. Consider categorizing your spending and setting monthly budgets for different categories like food, transport, and entertainment."
        
        elif "goal" in prompt_lower or "save" in prompt_lower:
            return "Setting clear financial goals is excellent! I recommend the SMART approach - Specific, Measurable, Achievable, Relevant, and Time-bound goals. What specific goal would you like to work on?"
        
        else:
            return "I'm here to help with all your financial needs - portfolio management, expense tracking, goal planning, and investment advice. What specific area would you like to explore?"
    
    def _get_financial_system_message(self) -> str:
        """Get comprehensive system message for financial assistant"""
        return """
        You are FinVoice AI, India's most advanced financial assistant. Your expertise includes:

        FINANCIAL SERVICES:
        - Portfolio analysis and investment advice
        - Expense tracking and budget optimization
        - Goal-based financial planning
        - SIP and mutual fund recommendations
        - Tax planning and optimization
        - Insurance and retirement planning

        INDIAN CONTEXT:
        - Use Indian Rupees (₹) for all amounts
        - Understand Indian financial instruments (PPF, EPF, ELSS, NPS)
        - Knowledge of Indian tax laws (80C, 80D, etc.)
        - Familiar with Indian markets (NSE, BSE, Nifty, Sensex)

        CONVERSATION STYLE:
        - Professional yet friendly and approachable
        - Provide specific, actionable advice
        - Use simple language, avoid financial jargon
        - Show empathy for financial concerns
        - Ask clarifying questions when needed

        RESPONSE GUIDELINES:
        - Keep responses conversational for voice interaction
        - Limit responses to 150 words maximum
        - Include specific numbers and examples
        - End with a question to continue conversation
        - Always include risk disclaimers for investment advice
        """
    
    def _get_greeting_message(self, language: str) -> str:
        """Get localized greeting message"""
        greetings = {
            "english": "Hello! I'm your FinVoice assistant. I can help you manage your finances, track expenses, and provide investment advice. How can I assist you today?",
            "hindi": "नमस्ते! मैं आपका FinVoice सहायक हूं। मैं आपके वित्त प्रबंधन, खर्च ट्रैकिंग और निवेश सलाह में मदद कर सकता हूं। आज मैं आपकी कैसे सहायता कर सकता हूं?",
            "tamil": "வணக்கம்! நான் உங்கள் FinVoice உதவியாளர். நான் உங்கள் நிதி நிர்வாகம், செலவு கண்காணிப்பு மற்றும் முதலீட்டு ஆலோசனையில் உதவ முடியும். இன்று நான் உங்களுக்கு எப்படி உதவ முடியும்?"
        }
        return greetings.get(language, greetings["english"])
    
    def _get_goodbye_message(self, language: str) -> str:
        """Get localized goodbye message"""
        goodbyes = {
            "english": "Thank you for using FinVoice! Keep building your financial future. Have a great day!",
            "hindi": "FinVoice का उपयोग करने के लिए धन्यवाद! अपना वित्तीय भविष्य बनाते रहें। आपका दिन शुभ हो!",
            "tamil": "FinVoice ஐப் பயன்படுத்தியதற்கு நன்றி! உங்கள் நிதி எதிர்காலத்தை உருவாக்கிக்கொண்டே இருங்கள். நல்ல நாள் இருக்கட்டும்!"
        }
        return goodbyes.get(language, goodbyes["english"])
    
    def _get_financial_functions(self) -> List[Dict[str, Any]]:
        """Get available functions for the assistant"""
        return [
            {
                "name": "get_portfolio_summary",
                "description": "Get user's portfolio summary including current value and performance",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "user_id": {"type": "string", "description": "User identifier"}
                    },
                    "required": ["user_id"]
                }
            },
            {
                "name": "add_expense",
                "description": "Add a new expense entry",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "amount": {"type": "number", "description": "Expense amount"},
                        "description": {"type": "string", "description": "Expense description"},
                        "category": {"type": "string", "description": "Expense category"}
                    },
                    "required": ["amount", "description"]
                }
            },
            {
                "name": "get_goal_progress",
                "description": "Get progress on financial goals",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "goal_id": {"type": "string", "description": "Goal identifier"}
                    }
                }
            }
        ]
    
    async def get_call_status(self, call_id: str) -> Dict[str, Any]:
        """Get voice call status"""
        if not self.private_key:
            return {"error": "Vapi not configured"}
        
        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                response = await client.get(
                    f"{self.base_url}/call/{call_id}",
                    headers={"Authorization": f"Bearer {self.private_key}"}
                )
                
                if response.status_code == 200:
                    return response.json()
                else:
                    return {"error": f"Failed to get call status: {response.status_code}"}
                    
        except Exception as e:
            logger.error(f"Error getting call status: {e}")
            return {"error": str(e)}
    
    async def end_call(self, call_id: str) -> bool:
        """End voice call"""
        if not self.private_key:
            return False
        
        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                response = await client.post(
                    f"{self.base_url}/call/{call_id}/end",
                    headers={"Authorization": f"Bearer {self.private_key}"}
                )
                
                return response.status_code == 200
                
        except Exception as e:
            logger.error(f"Error ending call: {e}")
            return False
    
    def get_webhook_handler(self):
        """Get webhook handler for Vapi events"""
        
        async def handle_webhook(request_data: Dict[str, Any]) -> Dict[str, Any]:
            """Handle incoming Vapi webhook events"""
            
            try:
                message_type = request_data.get("message", {}).get("type")
                
                if message_type == "function-call":
                    return await self._handle_function_call(request_data)
                elif message_type == "transcript":
                    return await self._handle_transcript(request_data)
                elif message_type == "call-start":
                    return await self._handle_call_start(request_data)
                elif message_type == "call-end":
                    return await self._handle_call_end(request_data)
                
                return {"status": "ok"}
                
            except Exception as e:
                logger.error(f"Webhook handling failed: {e}")
                return {"status": "error", "message": str(e)}
        
        return handle_webhook
    
    async def _handle_function_call(self, request_data: Dict[str, Any]) -> Dict[str, Any]:
        """Handle function calls from assistant"""
        
        function_call = request_data.get("message", {}).get("functionCall", {})
        function_name = function_call.get("name")
        parameters = function_call.get("parameters", {})
        
        if function_name == "get_portfolio_summary":
            # Return mock portfolio data
            return {
                "result": {
                    "total_value": "₹7,50,000",
                    "day_change": "+₹12,500 (+1.69%)",
                    "top_holdings": "Reliance, TCS, HDFC Bank",
                    "performance": "Your portfolio is performing well with diversified holdings"
                }
            }
        
        elif function_name == "add_expense":
            amount = parameters.get("amount")
            description = parameters.get("description")
            return {
                "result": {
                    "message": f"Added expense: {description} - ₹{amount}",
                    "status": "success"
                }
            }
        
        elif function_name == "get_goal_progress":
            return {
                "result": {
                    "goal_name": "Emergency Fund",
                    "progress": "76% complete",
                    "current_amount": "₹3,80,000",
                    "target_amount": "₹5,00,000"
                }
            }
        
        return {"result": {"message": "Function executed successfully"}}
    
    async def _handle_transcript(self, request_data: Dict[str, Any]) -> Dict[str, Any]:
        """Handle conversation transcript"""
        transcript = request_data.get("message", {}).get("transcript")
        logger.info(f"Voice transcript: {transcript}")
        return {"status": "logged"}
    
    async def _handle_call_start(self, request_data: Dict[str, Any]) -> Dict[str, Any]:
        """Handle call start event"""
        call_id = request_data.get("message", {}).get("call", {}).get("id")
        logger.info(f"Voice call started: {call_id}")
        return {"status": "call_started"}
    
    async def _handle_call_end(self, request_data: Dict[str, Any]) -> Dict[str, Any]:
        """Handle call end event"""
        call_id = request_data.get("message", {}).get("call", {}).get("id")
        duration = request_data.get("message", {}).get("call", {}).get("duration")
        logger.info(f"Voice call ended: {call_id}, duration: {duration}s")
        return {"status": "call_ended"}

File: frontend/src/services/vapiVoiceService.ts
/**
 * Vapi AI Voice Integration Service for Frontend
 * Uses PUBLIC KEY ONLY for secure client-side operations
 */

const BASE_URL = 'http://localhost:8000/api/v1';

interface VapiConfig {
  public_key: string;
  assistant_id: string;
  supported_languages: string[];
  features: {
    voice_calls: boolean;
    real_time_conversation: boolean;
    multi_language: boolean;
    financial_context: boolean;
  };
  endpoints: {
    start_call: string;
    voice_chat: string;
    call_status: string;
    webhook: string;
  };
}

interface VoiceMessage {
  message: string;
  language?: string;
  user_id?: string;
}

interface VoiceResponse {
  success: boolean;
  text_response: string;
  voice_url?: string;
  assistant_id?: string;
  user_message?: string;
  transcription?: string;
  audio_response?: string;
  error?: string;
}

interface CallRequest {
  phone_number: string;
  language?: string;
}

interface CallResponse {
  success: boolean;
  call_id?: string;
  assistant_id?: string;
  error?: string;
}

class VapiVoiceService {
  private config: VapiConfig | null = null;
  private publicKey: string;

  constructor() {
    // Get public key from environment (safe for frontend)
    this.publicKey = process.env.REACT_APP_VAPI_PUBLIC_KEY || '584ce1e5-dc84-43b3-9ec1-40099b3c315f';
  }

  private async apiRequest<T>(endpoint: string, options?: RequestInit): Promise<T> {
    try {
      const response = await fetch(`${BASE_URL}${endpoint}`, {
        headers: {
          'Content-Type': 'application/json',
          ...options?.headers,
        },
        ...options,
      });

      if (!response.ok) {
        throw new Error(`API Error: ${response.status} ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error(`API request failed for ${endpoint}:`, error);
      throw error;
    }
  }

  /**
   * Initialize Vapi service by fetching public configuration from backend
   */
  async initialize(): Promise<boolean> {
    try {
      const response = await this.apiRequest<{ config: VapiConfig }>('/vapi/config');
      this.config = response.config;
      return true;
    } catch (error) {
      console.error('Failed to initialize Vapi service:', error);
      return false;
    }
  }

  /**
   * Send voice message and get intelligent response
   */
  async sendVoiceMessage(message: VoiceMessage): Promise<VoiceResponse> {
    try {
      if (!this.config) {
        await this.initialize();
      }

      const response = await this.apiRequest<VoiceResponse>('/vapi/voice-chat', {
        method: 'POST',
        body: JSON.stringify({
          message: message.message,
          language: message.language || 'english',
          user_id: message.user_id || 'demo_user'
        })
      });

      return response;
    } catch (error) {
      console.error('Voice message error:', error);
      return {
        success: false,
        text_response: 'Sorry, I encountered an error processing your voice message.',
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  }

  /**
   * Start a voice call (phone call integration)
   */
  async startVoiceCall(callRequest: CallRequest): Promise<CallResponse> {
    try {
      if (!this.config) {
        await this.initialize();
      }

      const response = await this.apiRequest<CallResponse>('/vapi/start-call', {
        method: 'POST',
        body: JSON.stringify({
          phone_number: callRequest.phone_number,
          language: callRequest.language || 'english'
        })
      });

      return response;
    } catch (error) {
      console.error('Start call error:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to start call'
      };
    }
  }

  /**
   * Get call status
   */
  async getCallStatus(callId: string): Promise<any> {
    try {
      return await this.apiRequest(`/vapi/call-status/${callId}`);
    } catch (error) {
      console.error('Get call status error:', error);
      return { error: 'Failed to get call status' };
    }
  }

  /**
   * End voice call
   */
  async endCall(callId: string): Promise<{ success: boolean; error?: string }> {
    try {
      const response = await this.apiRequest<{ success: boolean; message?: string }>(`/vapi/end-call/${callId}`, {
        method: 'POST'
      });
      return { success: response.success };
    } catch (error) {
      console.error('End call error:', error);
      return { 
        success: false, 
        error: error instanceof Error ? error.message : 'Failed to end call' 
      };
    }
  }

  /**
   * Get voice capabilities
   */
  async getCapabilities(): Promise<any> {
    try {
      return await this.apiRequest('/voice/capabilities');
    } catch (error) {
      console.error('Get capabilities error:', error);
      return { error: 'Failed to get capabilities' };
    }
  }

  /**
   * Get supported languages
   */
  getSupportedLanguages(): string[] {
    return this.config?.supported_languages || ['english', 'hindi', 'tamil', 'telugu'];
  }

  /**
   * Check if service is available
   */
  isAvailable(): boolean {
    return this.config !== null;
  }

  /**
   * Get public configuration (safe to expose)
   */
  getPublicConfig(): Partial<VapiConfig> | null {
    if (!this.config) return null;

    return {
      supported_languages: this.config.supported_languages,
      features: this.config.features,
      assistant_id: this.config.assistant_id
    };
  }

  /**
   * Process enhanced voice message with file upload
   */
  async processVoiceFile(
    audioFile: File, 
    language: string = 'en-IN',
    useVapi: boolean = true
  ): Promise<VoiceResponse> {
    try {
      console.log('🎤 VapiVoiceService.processVoiceFile called with:', {
        fileName: audioFile.name,
        fileSize: audioFile.size,
        fileType: audioFile.type,
        language,
        useVapi
      });

      const formData = new FormData();
      formData.append('audio', audioFile);
      formData.append('language', language);
      formData.append('user_id', 'demo_user');
      formData.append('use_vapi', useVapi ? 'true' : 'false');

      console.log('🎤 Making request to:', `${BASE_URL}/voice/enhanced-message`);

      const response = await fetch(`${BASE_URL}/voice/enhanced-message`, {
        method: 'POST',
        body: formData
      });

      console.log('🎤 Response status:', response.status, response.statusText);

      if (!response.ok) {
        const errorText = await response.text();
        console.error('🎤 Response error:', errorText);
        throw new Error(`HTTP error! status: ${response.status}, message: ${errorText}`);
      }

      const result = await response.json();
      console.log('🎤 Response data:', result);

      return result;
    } catch (error) {
      console.error('❌ Voice file processing error:', error);
      return {
        success: false,
        text_response: 'Sorry, I could not process your voice message.',
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  }

  /**
   * Text-to-speech conversion
   */
  async textToSpeech(text: string, language: string = 'en-IN'): Promise<{ success: boolean; audio_response?: string; error?: string }> {
    try {
      const response = await this.apiRequest<{ success: boolean; audio_response?: string; error?: string }>('/voice/text-to-speech', {
        method: 'POST',
        body: JSON.stringify({
          text: text,
          language: language
        })
      });

      return response;
    } catch (error) {
      console.error('Text-to-speech error:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'TTS failed'
      };
    }
  }
}

// Export singleton instance
export const vapiVoiceService = new VapiVoiceService();
export default vapiVoiceService;

// Export types for use in components
export type {
  VapiConfig,
  VoiceMessage,
  VoiceResponse,
  CallRequest,
  CallResponse
};
